src/core/heavyGraph.js

```diff
@@ -0,0 +1,122 @@
+// @flow
+import {
+  type Edge,
+  type Node,
+  type EdgeAddressT,
+  type NodeAddressT,
+  Graph,
+} from "./graph";
+import {
+  type NodeWeight,
+  type EdgeWeight,
+  type WeightsComparison,
+} from "./weights";
+
+/**
+ * An entry in the array returned by AddressModule.toParts()
+ */
+export type AddressPart = string;
+
+/**
+ * the core structure in the Node weight-resolution tree
+ */
+export type NodeWeightBranch = {
+  +weight?: NodeWeight,
+  +children: Map<AddressPart, NodeWeightBranch>,
+};
+
+/**
+ * the core structure in the Edge weight-resolution tree
+ */
+export type EdgeWeightBranch = {
+  +weight?: EdgeWeight,
+  +children: Map<AddressPart, EdgeWeightBranch>,
+};
+
+/**
+ * The HeavyGraph class is a replacement for the WeightedGraph type. This class
+ * eliminates the need give every node and edge that exists a weight. Rather,
+ * address prefixes can be assigned a weight and when a node or edge weight is
+ * queried, the longest prefix with a weight assigned will have its value
+ * returned.
```

Comment:
This is the wrong semantics for graph weights; we return the product of all of the weights along the address path chain. See:

https://github.com/sourcecred/sourcecred/blob/651db6de017cc74dfc85f16960144d0849fab105/src/core/algorithm/weightEvaluator.js#L10-L63